File: KenKenAndroid/src/com/anthonysottile/kenken/ui/CustomButton.java
package com.anthonysottile.kenken.ui;

import java.util.ArrayList;
import java.util.EventListener;
import java.util.EventObject;
import java.util.List;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.MotionEvent;
import android.view.View;

public class CustomButton extends View {

	private final static int DefaultSize = 28;
	private final static int TopMargin = 8;
	
	// #region Check Changed Event
	
	public class CheckChangedEvent extends EventObject {
		
		private static final long serialVersionUID = 5703235645865455757L;
		
		private boolean checked = false;
		public boolean getChecked() {
			return this.checked;
		}
		
		public CheckChangedEvent(Object object, boolean checked) {
			super(object);
			this.checked = checked;
		}
	}
	public interface CheckChangedListener extends EventListener {
		
		public void onCheckChanged(CheckChangedEvent event);
	}
	private List<CheckChangedListener> checkChangedListeners =
			new ArrayList<CheckChangedListener>();
	public void AddCheckChangedListener(CheckChangedListener listener) {
		this.checkChangedListeners.add(listener);
	}
	public void RemoveCheckChangedListener(CheckChangedListener listener) {
		this.checkChangedListeners.remove(listener);
	}
	public void ClearCheckChangedListeners() {
		this.checkChangedListeners.clear();
	}
	private void triggerCheckChanged() {
        CheckChangedEvent event = new CheckChangedEvent(this, this.checked);
        
        for (CheckChangedListener listener : this.checkChangedListeners) {
        	listener.onCheckChanged(event);
        }
	}
	
	// #endregion
	
	// #region Click Event
	
	public interface ClickListener extends EventListener {
		public void onClick(EventObject event);
	}
	
	private List<ClickListener> clickListeners = new ArrayList<ClickListener>();
	public void AddClickListener(ClickListener listener) {
		this.clickListeners.add(listener);
	}
	public void RemoveClickListener(ClickListener listener) {
		this.clickListeners.remove(listener);
	}
	public void ClearClickListeners() {
		this.clickListeners.clear();
	}
	private void triggerClick() {
		EventObject event = new EventObject(this);
		
		for (ClickListener listener : this.clickListeners) {
			listener.onClick(event);
		}
	}
	
	// #endregion
	
	private int value = 0;
	
	/**
	 * Returns the integer value of this button.
	 * 
	 * @return The value of the button.
	 */
	public int getValue() {
		return this.value;
	}
	
	/**
	 * Sets the integer value of this button.
	 * 
	 * @param value The value to set on the button.
	 */
	public void setValue(int value) {
		this.value = value;
	}
	
	private boolean enabled = false;
	
	/**
	 * Returns whether the button is enabled.
	 * 
	 * @return Whether the button is enabled.
	 */
	public boolean getEnabled() {
		return this.enabled;
	}
	
	/**
	 * Sets whether the button is enabled.
	 * 
	 * @param enabled Whether the button should be enabled.
	 */
	public void setEnabled(boolean enabled) {
		if (this.enabled != enabled) {
			this.enabled = enabled;
			
			// Need to redraw
			this.postInvalidate();
		}
	}
	
	private boolean hasLeftCurve = false;
	
	/**
	 * Returns whether the button has a left curve.
	 * 
	 * @return Whether the button has a left curve.
	 */
	public boolean getHasLeftCurve() {
		return this.hasLeftCurve;
	}
	
	/**
	 * Sets whether the button has a left curve.
	 * 
	 * @param hasLeftCurve Whether the button has a left curve.
	 */
	public void setHasLeftCurve(boolean hasLeftCurve) {
		if (this.hasLeftCurve != hasLeftCurve) {
			this.hasLeftCurve = hasLeftCurve;
			
			// Need to redraw
			this.postInvalidate();
		}
	}

	private boolean hasRightCurve = false;
	
	/**
	 * Returns whether the button has a right curve.
	 * 
	 * @return Whether the button has a right curve.
	 */
	public boolean getHasRightCurve() {
		return this.hasRightCurve;
	}
	
	/**
	 * Sets whether the button has a right curve.
	 * 
	 * @param hasRightCurve Whether the button has a right curve.
	 */
	public void setHasRightCurve(boolean hasRightCurve) {
		if (this.hasRightCurve != hasRightCurve) {
			this.hasRightCurve = hasRightCurve;
			
			// Need to redraw
			this.postInvalidate();
		}
	}
	
	private boolean isCheckable = false;
	
	/**
	 * Returns whether the button reacts to clicking by checking and unchecking.
	 * 
	 * @return Whether the button reacts to clicking by checking and unchecking. 
	 */
	public boolean getIsCheckable() {
		return this.isCheckable;
	}
	
	/**
	 * Sets whether the button reacts to clicking by checking and unchecking.
	 * 
	 * @param isCheckable Whether the button reacts to clicking by checking and unchecking.
	 */
	public void setIsCheckable(boolean isCheckable) {
		this.isCheckable = isCheckable;
	}

	private boolean checked = false;
	
	/**
	 * Returns whether the button is checked.
	 * 
	 * @return Whether the button is checked.
	 */
	public boolean getChecked() {
		return this.checked;
	}
	
	/**
	 * Sets whether the button is checked.  Note this triggers check changed.
	 * 
	 * @param checked Whether the button is checked.
	 */
	public void setChecked(boolean checked) {
		if (this.checked != checked) {
			this.checked = checked;
			
			this.triggerCheckChanged();
			
			// Need to redraw
			this.postInvalidate();
		}
	}
	
	/**
	 * Sets whether the button is checked.  Note this does not trigger check changed.
	 * 
	 * @param checked Whether the button is checked.
	 */
	public void setCheckedNoTrigger(boolean checked) {
		if (this.checked != checked) {
			this.checked = checked;
			
			// Need to redraw
			this.postInvalidate();
		}
	}
	
	/**
	 * Toggles the checked state of the button. Note this triggers check changed.
	 */
	public void toggleChecked() {
		this.setChecked(!this.checked);
	}
	
	private String text = "";
	
	/**
	 * Returns the text of the button.
	 * 
	 * @return The text of the button.
	 */
	public String getText() {
		return this.text;
	}
	
	/**
	 * Sets the text of the button.
	 * 
	 * @param text The display text of the button.
	 */
	public void setText(String text) {
		if (!this.text.equals(text)) {
			this.text = text;
			
			// Redraw
			this.postInvalidate();
		}
	}
	
	@Override
	public boolean onTouchEvent(MotionEvent event) {
		
		// Click event
		if (this.enabled) {
		
			float x = event.getX();
			float y = event.getY();
			
			// Determine if the click position is inside the control
			boolean isInsideControl =
				(x >= 0 && x <= this.getMeasuredWidth())
				&& (y >= 0 && y <= this.getMeasuredHeight());
			
			switch(event.getAction()) {
				case MotionEvent.ACTION_DOWN:
				case MotionEvent.ACTION_MOVE:
					
					break;
				case MotionEvent.ACTION_UP:
					
					if (isInsideControl) {
						
						if (this.isCheckable) {
							// If we are checkable then change the checked state
							this.setChecked(!this.checked);
						}
						
						this.triggerClick();
					}
					
					break;
			}
		}

		return this.enabled;
	}
	
	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int widthMode = MeasureSpec.getMode(widthMeasureSpec);
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int heightMode = MeasureSpec.getMode(heightMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);
		
        int width = -1, height = -1;
        if (widthMode == MeasureSpec.EXACTLY) {
        	width = widthSize;
        } else {
        	width = CustomButton.DefaultSize;
        	if (widthMode == MeasureSpec.AT_MOST && width > widthSize ) {
        		width = widthSize;
        	}
        }
        if (heightMode == MeasureSpec.EXACTLY) {
        	height = heightSize;
        } else {
        	height = CustomButton.DefaultSize;
        	if (heightMode == MeasureSpec.AT_MOST && height > heightSize ) {
        		height = heightSize;
        	}
        }
        
		this.setMeasuredDimension(width, height);
	}
	
	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
		
		int start = 1;
		int end = this.getMeasuredWidth() - 1;

		Paint p = UIConstants.GetCageColor();
		
		// Draw left curve
		if (this.hasLeftCurve) {
			start += 7;
			Rect destRect = new Rect(0, 0, 8, this.getMeasuredHeight());
			
			Bitmap bitmap;
			if (!this.enabled) {
				bitmap = BitmapCache.getDisabledLeft();
			} else if (this.checked) {
				bitmap = BitmapCache.getSelectedLeft();
			} else {
				bitmap = BitmapCache.getEnabledLeft();
			}
			
			// Draw left curve
			canvas.drawBitmap(bitmap, null, destRect, p);
				
		} else {
			// Draw 1px left border
			canvas.drawLine(0, 0, 0, this.getMeasuredHeight(), p);
		}
		
		// Draw right curve
		if (this.hasRightCurve) {
			end -= 7;
			Rect destReg = new Rect(
				this.getMeasuredWidth() - 8,
				0,
				this.getMeasuredWidth(),
				this.getMeasuredHeight()
			);
			
			Bitmap bitmap;
			if (!this.enabled) {
				bitmap = BitmapCache.getDisabledRight();
			} else if (this.checked) {
				bitmap = BitmapCache.getSelectedRight();
			} else {
				bitmap = BitmapCache.getEnabledRight();
			}
			
			// Draw right curve
			canvas.drawBitmap(bitmap, null, destReg, p);
			
		} else {
			// Draw 1px right border
			canvas.drawLine(
				this.getMeasuredWidth() - 1,
				0,
				this.getMeasuredWidth() - 1,
				this.getMeasuredHeight(),
				p
			);
		}
		
		// Draw the background
		
		Bitmap bitmap;
		if (!this.enabled) {
			bitmap = BitmapCache.getDisabledCenter();
		} else if (this.checked) {
			bitmap = BitmapCache.getSelectedCenter();
		} else {
			bitmap = BitmapCache.getEnabledCenter();
		}
		
		Rect destReg = new Rect(
			start,
			0,
			end,
			this.getMeasuredHeight()
		);
		
		canvas.drawBitmap(bitmap, null, destReg, p);
		
		// Paint the text
		
		int textSize = this.getMeasuredHeight() - CustomButton.TopMargin * 2;
		
		Paint textPaint = new Paint();
		textPaint.setTextSize(textSize);
		if (!this.enabled) {
			textPaint.setColor(Color.rgb(0x99, 0x99, 0x99));
		} else if (checked) {
			textPaint.setColor(Color.rgb(0, 0, 0));
		} else {
			textPaint.setColor(Color.rgb(0xff, 0xff, 0xff));
		}
		
		float textWidth = textPaint.measureText(this.text);
		int left = (this.getMeasuredWidth() - (int)textWidth) / 2;
		int top = CustomButton.TopMargin + textSize;
		
		canvas.drawText(this.text, left, top, textPaint);
	}
	
	public CustomButton(Context context) {
		super(context);
		
		this.postInvalidate();
	}	
}

File: app/src/main/java/com/anthonysottile/kenken/ui/CustomButton.kt
package com.anthonysottile.kenken.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.view.MotionEvent
import android.view.View
import java.util.*

class CustomButton(
        context: Context,
        uiEnabled: Boolean,
        checked: Boolean,
        val value: Int,
        val text: String,
        private val hasLeftCurve: Boolean,
        private val hasRightCurve: Boolean,
        private val isCheckable: Boolean
) : View(context) {
    private val destRect = Rect()
    private val textPaint = Paint()

    private val checkChangedListeners = ArrayList<(obj: CustomButton) -> Unit>()
    private val clickListeners = ArrayList<(obj: CustomButton) -> Unit>()

    var uiEnabled = uiEnabled
        set(enabled) {
            if (field != enabled) {
                field = enabled
                this.postInvalidate()
            }
        }

    private var checkedVal: Boolean = checked

    var checked: Boolean
        get() = this.checkedVal
        set(checked) {
            if (this.checkedVal != checked) {
                this.checkedVal = checked
                this.triggerCheckChanged()
                this.postInvalidate()
            }
        }

    fun setCheckedNoTrigger(checked: Boolean) {
        if (this.checkedVal != checked) {
            this.checkedVal = checked
            this.postInvalidate()
        }
    }

    fun addCheckChangedListener(listener: (obj: CustomButton) -> Unit) {
        this.checkChangedListeners.add(listener)
    }

    private fun triggerCheckChanged() {
        for (listener in this.checkChangedListeners) {
            listener(this)
        }
    }

    fun addClickListener(listener: (obj: CustomButton) -> Unit) {
        this.clickListeners.add(listener)
    }

    private fun triggerClick() {
        for (listener in this.clickListeners) {
            listener(this)
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        // Click event
        if (this.uiEnabled) {
            val x = event.x
            val y = event.y

            // Determine if the click position is inside the control
            val isInsideControl = x >= 0 && x <= this.measuredWidth && y >= 0 && y <= this.measuredHeight

            when (event.action) {
                MotionEvent.ACTION_UP ->
                    if (isInsideControl) {
                        if (this.isCheckable) {
                            // If we are checkable then change the checked state
                            this.checked = !this.checked
                        }

                        this.triggerClick()
                    }
            }
        }

        return this.uiEnabled
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val widthMode = View.MeasureSpec.getMode(widthMeasureSpec)
        val widthSize = View.MeasureSpec.getSize(widthMeasureSpec)
        val heightMode = View.MeasureSpec.getMode(heightMeasureSpec)
        val heightSize = View.MeasureSpec.getSize(heightMeasureSpec)

        var width: Int
        var height: Int
        if (widthMode == View.MeasureSpec.EXACTLY) {
            width = widthSize
        } else {
            width = CustomButton.DefaultSize
            if (widthMode == View.MeasureSpec.AT_MOST && width > widthSize) {
                width = widthSize
            }
        }
        if (heightMode == View.MeasureSpec.EXACTLY) {
            height = heightSize
        } else {
            height = CustomButton.DefaultSize
            if (heightMode == View.MeasureSpec.AT_MOST && height > heightSize) {
                height = heightSize
            }
        }

        this.setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        var start = 1
        var end = this.measuredWidth - 1

        // Draw left curve
        if (this.hasLeftCurve) {
            start += 7
            this.destRect.set(0, 0, 8, this.measuredHeight)

            val bitmap = if (!this.uiEnabled) {
                BitmapCache.disabledLeft
            } else if (this.checked) {
                BitmapCache.selectedLeft
            } else {
                BitmapCache.enabledLeft
            }

            // Draw left curve
            canvas.drawBitmap(bitmap, null, this.destRect, paint)

        } else {
            // Draw 1px left border
            canvas.drawLine(0f, 0f, 0f, this.measuredHeight.toFloat(), paint)
        }

        // Draw right curve
        if (this.hasRightCurve) {
            end -= 7
            this.destRect.set(
                    this.measuredWidth - 8,
                    0,
                    this.measuredWidth,
                    this.measuredHeight
            )

            val bitmap = if (!this.uiEnabled) {
                BitmapCache.disabledRight
            } else if (this.checked) {
                BitmapCache.selectedRight
            } else {
                BitmapCache.enabledRight
            }

            // Draw right curve
            canvas.drawBitmap(bitmap, null, this.destRect, paint)

        } else {
            // Draw 1px right border
            canvas.drawLine(
                    (this.measuredWidth - 1).toFloat(),
                    0f,
                    (this.measuredWidth - 1).toFloat(),
                    this.measuredHeight.toFloat(),
                    paint
            )
        }

        // Draw the background
        val bitmap = if (!this.uiEnabled) {
            BitmapCache.disabledCenter
        } else if (this.checked) {
            BitmapCache.selectedCenter
        } else {
            BitmapCache.enabledCenter
        }

        this.destRect.set(
                start,
                0,
                end,
                this.measuredHeight
        )

        canvas.drawBitmap(bitmap, null, this.destRect, paint)

        // Paint the text

        val textSize = this.measuredHeight - CustomButton.TopMargin * 2

        this.textPaint.textSize = textSize.toFloat()
        if (!this.uiEnabled) {
            this.textPaint.color = Color.rgb(0x99, 0x99, 0x99)
        } else if (this.checked) {
            this.textPaint.color = Color.rgb(0, 0, 0)
        } else {
            this.textPaint.color = Color.rgb(0xff, 0xff, 0xff)
        }

        val textWidth = this.textPaint.measureText(this.text)
        val left = (this.measuredWidth - textWidth.toInt()) / 2
        val top = CustomButton.TopMargin + textSize

        canvas.drawText(this.text, left.toFloat(), top.toFloat(), this.textPaint)
    }

    companion object {
        private const val DefaultSize = 28
        private const val TopMargin = 8
        private val paint = Paint()
    }
}
